### PR TITLE
Migrated listAssertions to atrium-api-infix-en_GB-common

### LINKED ISSUE #228: listAssertions for the new infix API
*Platform* (jvm, js, android): jvm
*Extension* (none, kotlin 1.3, jdk8): none

## Code related list

We want to migrate the `api-cc-infix-en_GB...` modules which are based on `Assert` to `Expect`. We have already created the corresponding
modules `api-infix-en_GB...`. Also there are already `fluent-en_GB...` modules where we have already done the transition from `Assert` to `Expect`

Following the things you need to do:

- copy the file `listAssertions.kt` from atrium-api-fluent-en_GB-common to atrium-api-infix-en_GB-common
- copy the `ListFeatureAssertionsSpec.kt` from atrium-api-fluent-en_GB-common to atrium-api-infix-en_GB-common
  - introduce a function in the companion object for each assertion function so that we see the infix API in action  (see ListFeatureAssertionsSpec in atrium-api-cc-infix-en_GB-jvm but use fun1, fun2 etc. to retrieve the name of the function instead of `KFunction<...>` -- use `fun1<..>(...).name`)

## Your first contribution?
- Write a comment `I'll work on this` if you would like to take this issue over. 
  This way we get the chance to revise the description in case things have changed in the meantime,
  we might give you additional hints and we can assign the task to you, so that others do not start as well.
- See [Your first code contribution](https://github.com/robstoll/atrium/blob/master/.github/CONTRIBUTING.md#your-first-code-contribution) for guidelines.  
- Do not hesitate to ask questions here or to contact us via [Atrium's slack channel](https://kotlinlang.slack.com/team/U3DE1TXKP) if you need help
  ([Invite yourself](https://slack.kotlinlang.org/) in case you do not have an account yet).

### PR DESCRIPTION
**Pull request by sanatik**
Fixes #228 
----
I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/master/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

### CODE REVIEW

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/creating/list/get/builders/ListGetStep.kt [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package ch.tutteli.atrium.api.infix.en_GB.creating.list.get.builders
+
+import ch.tutteli.atrium.api.infix.en_GB.creating.list.get.builders.impl.ListGetStepImpl
+import ch.tutteli.atrium.assertions.Assertion
+import ch.tutteli.atrium.creating.Expect
+import ch.tutteli.atrium.creating.SubjectProvider
+
+/**
+ * Represents the extension point for another option after a `get index`-step within a
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/creating/list/get/builders/ListGetStep.kt]:
```suggestion
 * Represents the extension point for another step after a `get index`-step within a
```

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/creating/list/get/builders/ListGetStep.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * sophisticated `get` assertion building process for [List].
+ *
+ * @param E The element type of the [List].
+ * @param T A subtype of [List].
+ */
+interface ListGetStep<E, T: List<E>> {
+    /**
+     * The [Expect] for which this assertion is created
+     */
+    val plant: Expect<T>
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/creating/list/get/builders/ListGetStep.kt]:
```suggestion
    val asertionContainer: Expect<T>
```

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/creating/list/get/builders/ListGetStep.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * The given index which will be used to perform the [List.get].
+     */
+    val index: Int
+
+    /**
+     * Makes the assertion that the given [index] is within the bounds of [Expect.subject][SubjectProvider.subject] and that
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/creating/list/get/builders/ListGetStep.kt]:
```suggestion
     * Makes the assertion that the given [index] is within the bounds of [Expect.subject] and that
```

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/creating/list/get/builders/ListGetStep.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+     *
+     * @return This plant to support a fluent API.
+     * @throws AssertionError Might throw an [AssertionError] if a created [Expect]s (by calling [assertionCreator])
+     *   does not hold.
+     * @throws IllegalArgumentException in case the given [assertionCreator] did not create a single assertion.
+     */
+    infix fun assertIt(assertionCreator: Expect<E>.() -> Unit): Expect<T>
+
+    companion object {
+        fun <E, T: List<E>> create(plant: Expect<T>, index: Int): ListGetStep<E, T>
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/creating/list/get/builders/ListGetStep.kt]:
```suggestion
        fun <E, T: List<E>> create(assertionContainer: Expect<T>, index: Int): ListGetStep<E, T>
```

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/creating/list/get/builders/ListGetStep.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+            = ListGetStepImpl(plant, index)
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/creating/list/get/builders/ListGetStep.kt]:
```suggestion
            = ListGetStepImpl(assertionContainer, index)
```

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/creating/list/get/builders/impl/ListGetStepImpl.kt [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package ch.tutteli.atrium.api.infix.en_GB.creating.list.get.builders.impl
+
+import ch.tutteli.atrium.api.infix.en_GB.creating.list.get.builders.ListGetStep
+import ch.tutteli.atrium.creating.Expect
+import ch.tutteli.atrium.domain.builders.ExpectImpl
+
+internal class ListGetStepImpl<E, T: List<E>>(
+    override val plant: Expect<T>,
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/creating/list/get/builders/impl/ListGetStepImpl.kt]:
```suggestion
    override val assertionContainer: Expect<T>,
```

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/creating/list/get/builders/impl/ListGetStepImpl.kt [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    override val index: Int
+) : ListGetStep<E, T> {
+
+    override infix fun assertIt(assertionCreator: Expect<E>.() -> Unit): Expect<T>
+        = plant.addAssertion(ExpectImpl.list.get(plant, index).collect(assertionCreator))
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/creating/list/get/builders/impl/ListGetStepImpl.kt]:
```suggestion
        = assertionContainer.addAssertion(ExpectImpl.list.get(assertionContainer, index).collect(assertionCreator))
```

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/listAssertions.kt [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+import ch.tutteli.atrium.domain.builders.ExpectImpl
+
+/**
+ * Expects that the given [index] is within the bounds of the subject of the assertion (a [List]) and
+ * returns an [Expect] for the element at that position.
+ *
+ * @return The newly created [Expect] for the element at position [index].
+ * @throws AssertionError if the given [index] is out of bound.
+ *
+ * @since 0.9.0
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/listAssertions.kt]:
```suggestion

```
This functionality kind of existed in 0.8.0 as well, just in another API module. Therefore we don't add it.

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/listAssertions.kt [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+ * Expects that the given [index] is within the bounds of the subject of the assertion (a [List]) and that
+ * the element at that position holds all assertions the given [assertionCreator] creates for it.
+ *
+ * @return This assertion container to support an infix API.
+ * @throws AssertionError if the given [index] is out of bound.
+ *
+ * @since 0.9.0
+ */
+fun <E, T : List<E>> Expect<T>.get(index: Int, assertionCreator: Expect<E>.() -> Unit): Expect<T> =
+    ExpectImpl.list.get(this, index).addToInitial(assertionCreator)
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/listAssertions.kt]: remove this one, this is covered by the next function using `Index` and `ListGetStep`

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/listAssertions.kt [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * Prepares the assertion about the return value of calling [get][List.get] with the given [index].
+ *
+ * @return A fluent builder to finish the assertion.
+ *
+ * @since 0.9.0
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/infix/en_GB/listAssertions.kt]:
```suggestion

```

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+            getPlant(a1, 1)
+            getPlant(a1, Index(1))
+            a1 = a1.get(1) { }
+            getPlant(a2, 1)
+            getPlant(a2, Index(1))
+            a2 = a2.get(1) { }
+        }
+
+        private fun getPlant(plant: Expect<out List<Int>>, index: Int) = plant get index
+        private fun getPlant(plant: Expect<out List<Int>>, index: Index) = plant get index
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt]: The ambiguity test shall reveal bugs in the implementation such as `Expect<Iterable<*>>.hasNext()` does not work. However, for this we need to duplicate code, therefore, please remove this function. Please rebase on master, I have just improved the ambiguityTests in atrium-api-fluent-en_GB. Copy the newer version.

Comment (sanatik) [reply apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt]: Not sure what function exactly you ask to remove.
In issue description there was a point, where you asked to check "real" infix function usage and this `getPlant` function is doing it for different indices.
You want me to remove both?
How ambiguity test will look like for me then? Is it something like this then?
```
        a1.get(1)
        a2.get(1)
        a1b.get(1)
        a2b.get(1)
        a3.get(1)
```

Comment (robstoll) [reply apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt]: Both getPlant functions. Use the you API defined directly. Use tge overload which expects `Index` for the case with the assertionCreator

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+    companion object {
+
+        @Suppress("unused", "UNUSED_VALUE")
+        private fun ambiguityTest() {
+
+            val a1: Expect<List<Int>> = notImplemented()
+            val a2: Expect<out List<Int>> = notImplemented()
+            val a1b: Expect<List<Int?>> = notImplemented()
+            val a2b: Expect<out List<Int?>> = notImplemented()
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt]: turn into `var`

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+            a1 get 1
+            a2 get 1
+
+            a1b get 1
+            a2b get 1
+
+            a3 get 1
+
+            a1 get Index(1)
+            a2 get Index(1)
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt]: reassign variables (see ch.tutteli.atrium.api.fluent.en_GB.ListFeatureAssertionsSpec) which means you need to finish the ListGetStep with `assertIt {}`

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+        private fun ambiguityTest() {
+
+            val a1: Expect<List<Int>> = notImplemented()
+            val a2: Expect<out List<Int>> = notImplemented()
+            val a1b: Expect<List<Int?>> = notImplemented()
+            val a2b: Expect<out List<Int?>> = notImplemented()
+
+            val a3: Expect<out List<*>> = notImplemented()
+
+            a1 get 1
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt]: please change the structure, first a1 and a2 then a1b and a2b and finally a3 (see ch.tutteli.atrium.api.fluent.en_GB.ListFeatureAssertionsSpec) this should make it easier to add further functionality in an ordered manner

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt [hunk 15]
```diff
@@ -17,28 +17,30 @@ class ListFeatureAssertionsSpec : ch.tutteli.atrium.specs.integration.ListFeatur
...
             a1 get 1
             a2 get 1
-
-            a1b get 1
-            a2b get 1
-
-            a3 get 1
-
             a1 get Index(1)
             a2 get Index(1)
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt]: no need for those then, covered by the next two lines

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt [hunk 16]
```diff
@@ -17,28 +17,30 @@ class ListFeatureAssertionsSpec : ch.tutteli.atrium.specs.integration.ListFeatur
...
+            a1 = a1 get Index(1) assertIt { }
+            a2 = a2 get Index(1) assertIt { }
 
+            a1b get 1
+            a2b get 1
             a1b get Index(1)
             a2b get Index(1)
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt]: same here

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+            a2 get 1
+            a1 get Index(1)
+            a2 get Index(1)
+            a1 = a1 get Index(1) assertIt { }
+            a2 = a2 get Index(1) assertIt { }
+
+            a1b get 1
+            a2b get 1
+            a1b get Index(1)
+            a2b get Index(1)
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt]: no need for those then, covered by the next two lines

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+            a1b = a1b get Index(1) assertIt { }
+            a2b = a2b get Index(1) assertIt { }
+
+            a3 get 1
+            a3 get Index(1)
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt]: same here

File: apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+            var a2: Expect<out List<Int>> = notImplemented()
+            var a1b: Expect<List<Int?>> = notImplemented()
+            var a2b: Expect<out List<Int?>> = notImplemented()
+
+            var a3: Expect<out List<*>> = notImplemented()
+
+            a1 get 1
+            a2 get 1
+            a1 get Index(1)
+            a2 get Index(1)
```
Comment (robstoll) [inline apis/infix-en_GB/atrium-api-infix-en_GB-common/src/test/kotlin/ch/tutteli/atrium/api/infix/en_GB/ListFeatureAssertionsSpec.kt]: no need for those then, covered by the next two lines

**Comment by robstoll** [comment]
@sanatik thanks also for this contribution :slightly_smiling_face: :+1:

**Comment by robstoll** [comment]
@sanatik Oh well, I merged a failing PR, that never happened to me before :D
I must have been tired, I also did not realise that you haven't used the infix API for the spec. 
The following fixed the build: 4cab4ad94e021629817fa4da6fee9201620005b9

**Comment by sanatik** [comment]
@robstoll I realized it when I tried to run the test after you requested latest changes.
Then I was trying to fix them and you merged the PR :D 
Sorry for that, testing infix functions is a bit harder than I thought.

**Comment by robstoll** [comment]
no worries, should have waited until the build became green

**Pull request merged**